### PR TITLE
feat(clustering): Phase 2 UMAP tuning — dry-run, gate metrics, findings template (#327)

### PR DESCRIPTION
**Pull request by BryanPMX**
## Summary

Phase 2 of JIE #327 — adds the tooling to execute the one-variable-at-a-time UMAP tuning sweep safely (no prod DB writes) and generate structured Markdown findings notes automatically.

Contributes to #327 (Phase 2 tooling; actual tuning runs + findings docs are committed as the sweep progresses).

### `scripts/run_clustering.py`

| Addition | Purpose |
|----------|---------|
| `--dry-run` flag | Skips DB persistence — safe to run tuning iterations without touching `canonical_roles` or `role_snapshot_weekly` |
| `--findings-output PATH` | Writes a Markdown findings doc with config snapshot, gate verdict, and top-cluster table |
| Mega-cluster share calculation | `mega_cluster_share = top_cluster.member_count / eligible_posting_count` — required for Phase 2 gate |
| Phase 2 gate inline verdict | Prints PASS/FAIL against `clusters≥20, mega<10%, noise<30%` on every run |
| `config_snapshot` in findings JSON | Captures the exact 7 HDBSCAN+UMAP params for reproducibility |
| Module-level docstring | Documents the Phase 2 tuning matrix (priority 1–4) with env-override commands |

### `eval/runs/findingsv2.phase2-clustering-template.md`

Template for each Phase 2 tuning iteration's findings note, covering:
- Full parameter snapshot (what changed vs prior run)
- Gate table with pass/fail per metric
- Qualitative cluster-label review prompt (5 sample postings per cluster)
- Semantic coherence check + decision section

## How to run Phase 2 (one variable at a time)

```bash
# Priority 1: min_samples sweep (current = 3; try 2 and 5)
CLUSTER_MIN_SAMPLES=2 python scripts/run_clustering.py \
    --dry-run \
    --findings-output eval/runs/findingsv2.1-clustering-tier3-min-samples-2.md

CLUSTER_MIN_SAMPLES=5 python scripts/run_clustering.py \
    --dry-run \
    --findings-output eval/runs/findingsv2.1-clustering-tier3-min-samples-5.md

# Priority 2: min_cluster_size sweep (try 2 and 5)
CLUSTER_MIN_CLUSTER_SIZE=2 python scripts/run_clustering.py --dry-run ...
```

After each run: fill in the findings template, update `config/clustering.yaml` with the winner.

## Test plan

- [ ] `python scripts/run_clustering.py --help` shows `--dry-run` and `--findings-output`
- [ ] `--dry-run` run completes without DB write (verify `canonical_roles` count unchanged)
- [ ] `--findings-output` produces a valid Markdown file with gate verdict and cluster table
- [ ] Phase 2 priority 1: run `CLUSTER_MIN_SAMPLES=2` and `CLUSTER_MIN_SAMPLES=5`; fill findings docs; commit winner config

## Phase 2 tuning matrix (from #327)

| Priority | Knob | Current → Try | Goal |
|----------|------|---------------|------|
| 1 | `min_samples` | 3 → 2, 5 | More clusters survive |
| 2 | `min_cluster_size` | 3 → 2, 5 | More candidate clusters |
| 3 | UMAP `n_components` | 15 → 10, 20 | More semantic separation |
| 4 | UMAP `n_neighbors` | 10 → 5, 15 | Fine-tune cluster locality |

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Review by Milol0608 — CHANGES_REQUESTED** [review]
## PR Review — #415: feat(clustering): Phase 2 UMAP tuning — dry-run, gate metrics, findings template (#327)

**PR:** https://github.com/Building-With-Agents/job-intelligence-engine/pull/415  
**Branch:** `feat/327-phase2-umap-tuning` → `development`  
**Author:** BryanPMX  
**Status:** Draft  
**Issues:** #327 (contributes tooling; does not close)  
**Recommendation:** 🔄 Request changes

### Summary

Adds Phase 2 clustering tuning tooling: `--dry-run` and `--findings-output` on `scripts/run_clustering.py`, mega-cluster share / gate verdict printing, and a findings Markdown template. Useful for safe one-variable-at-a-time UMAP sweeps. CI lint fails on ruff format; no dedicated tests for new flags. Still requires a live DB even in dry-run mode (skips writes only).

### Issues addressed

| Issue | Title | Verdict | Evidence |
|-------|-------|---------|----------|
| #327 | Phase 2 UMAP tuning | ⚠️ Partial | Tooling only; actual tuning runs + committed findings not in this PR |

### Scope & correctness

- **In scope:** Clustering script enhancements + eval runbook template for #327 Phase 2 sweep.
- **Phase 2 note:** This is issue #327 "Phase 2 tuning" (clustering hyperparams), not the nine-agent Phase 2 backlog — acceptable for analytics tooling.
- **PR claims vs implementation:** Matches PR body; `--dry-run` means no DB writes, not offline/no-DB.

### Tests & CI

| Check | Result | Notes |
|-------|--------|-------|
| GitHub CI lint | FAIL | `scripts/run_clustering.py` — ruff format |
| GitHub CI test | PASS | |
| Dedicated `--dry-run` / `--findings-output` tests | ❌ Missing | No `analytics/clustering/tests/` for script flags |

### Findings

#### 🔴 Blockers

- **CI lint red:** Run `ruff format scripts/run_clustering.py` (CI ruff 0.15.15 failed; local Windows ruff may disagree — verify in Linux/CI-equivalent env).

#### 🟡 Suggestions

- Add a lightweight unit test for gate verdict logic / findings Markdown generation (mock clustering result).
- Document that `--dry-run` still needs `PYTHON_DATABASE_URL` and feature load.
- DRY: gate thresholds appear in stdout, JSON, and `_write_findings_md` — extract to named constants.
